### ISSUE TITLE
Not showing a validation message for conditional questions with the recent ver. of schema-form.js

### ISSUE THREAD

**Issue by warmy1004** [json-schema-form/angular-schema-form#649]
Hello all,

I have an issue to show a validation message for conditional questions.  Just in case, I am using the recent version of schema-form.js which is updated by Angular Schema form team 6 months ago. The version number of this schema-form.js in their package.json is 0.8.12.

Here is my example: https://plnkr.co/edit/GpmbaMPZmtgofOdV0Fye?p=preview
How to reproduce the issue in my example:
(1) select 'YES' for showing some conditional questions from the first question
      The first question is "Do you want to see the 2nd questions?".
(2) just click 'Validate' button to validate the entire form.
(3) You can see some css changes (color change to red), but you can't see any error message (sf-message) under each field, which should be 'Required'. 

However, if you click 'validate' button for the 1st question without any selected value(Yes or No), you can see a proper error message (Required) because this is not the conditional questions.
how to see this: 
(1) Just click 'validate' button from the first question without any selected value.
(2) Then, you can see "Required" red error message under the radio buttons.

Do you have any idea for this?
Thank you,

**Comment by warmy1004** [comment]
Previously, I was using an old version of schema-form.js. So, I updated the proper recent schema-form.js which I tested.

**Comment by warmy1004** [comment]
FYI, watchCollection in sfMessage is not called for the conditional questions through $scope.$broadcast('schemaFormValidate').

**Comment by Anthropic** [comment]
@warmy1004 was the result the same with the latest version?

**Comment by warmy1004** [comment]
@Anthropic In the above plunker, I used v.0.8.12 which is the recent version as far as I know. And, with the previous version of schema-form instead of 0.8.12, the result was different. FYI, the issue with the previous version was the shown error message for radios-inline question didn't disappear when I selected 'yes' first. But, it worked fine if I selected 'No' first. Unfortunately, I don't know what the version of the previous one exactly is.

**Comment by Anthropic** [comment]
@warmy1004 thank you so much for the plunkers they are a great help, I'm not sure if this will be looked at before or after the separation of the core library is completed, but I will keep it on the radar to be looked into asap.

**Comment by Anthropic** [comment]
I'm about to release 0.8.13, curious to see if that has any affect.

**Comment by warmy1004** [comment]
@Anthropic Okay. Thank you. BTW, can you let me know when the next release 0.8.13 will come?

**Comment by Anthropic** [comment]
@warmy1004 I'm hoping for today or tomorrow, it's already in the master branch, I'm just about to tag it and then get David to push to npm, but in theory it should be on bower as soon as it is tagged.

**Comment by warmy1004** [comment]
@Anthropic Just fyi, I downloaded 0.8.13 version and tested it. But, my issue is not solved as we expected. Thank you.

**Comment by warmy1004** [comment]
Hello, I want to share what I found for solving this issue.
In 0.8.12 version of Schema-from, the conditional questions' missing error message text issue is caused by validation when the conditional questions are loaded. That is, when the conditional questions display, schema-form validates the forms, and those forms are set to have 'tv4-302' (required) error even though they are still pristine. So, when the actual 'tv4-302' validation happens, ngModel.$error for the conditional questions are not changed and the below $watchCollection
      scope.$watchCollection('ngModel.$error', function() {
              update(!!scope.ngModel);
      }); 
can't be run for calling setMessage(). So, I changed one line inside schemaValidate directive's validate() from _if (!result.valid) {_ to _if (!result.valid && !ngModel.$pristine) {_. With this change, the conditional question issue looks like fixed.

**Comment by Anthropic** [comment]
@warmy1004 that's great thanks, can you make a PR or let me know which file and line in the src it is?

**Comment by warmy1004** [comment]
@Anthropic I don't know how to make a PR, so I'll just let you know a file name and line number. 
The changed file is /src/directives/schema-validate.js and line number is 52. So, a current code on line 52 which is "_if(!result.valid) {_" needs to be changed to "**_if(!result.valid && $ngModel.$pristine){_**" for both v 0.8.12 and v 0.8.13.

**Comment by Anthropic** [comment]
@warmy1004 thanks I will take a look :)

FYI to make a PR you fork the repo (using the fork button top right), pull it down locally to work on, make changes in your forked version and then from your forked version you can then request your changes be pulled from your repo into this one, if you made the changes in a fork I could guide you through the rest if you are ever interested.

Thanks for your time,
Marcel

**Comment by warmy1004** [comment]
@Anthropic Oh I see. Thank you so much for your explanation :)

**Comment by Anthropic** [comment]
@warmy1004 can you give this a try in alpha.3?

**Comment by warmy1004** [comment]
@Anthropic I tested it with alpha.3, but it's not working as expected. And it seems like there is another issue; it looks like validation is always recursively and automatically happening for the entire form including not displayed things on the screen. So, the parent question is getting 'Required' message even though conditional questions are not displayed. 

To help your understanding, follow this step to reproduce it.
1. select 'yes' for the parent question (Do you want to see the 2nd questions?) 
2. then, you can see three conditional questions. 
3. then, switch your answer of the parent question from 'yes' to 'No'
-> Then, you can see 'Required' error message and the parent question is in error state. 
Broadcast for schemaFormValidate only happens when 'validate' button is clicked. In this flow, you didn't click this button, so the expected result after step 3 is just hiding three conditional questions and the parent question should not be in error state. 

FYI, I checked this validation issue with version 0.8.9, this version is working fine. But, 0.8.12, 0.8.13, and alpha.3 are not working as expected. 

If you have any question, let me know please.

**Comment by Anthropic** [comment]
@warmy1004 can you try again with the matching bootstrap decorator:
https://raw.githubusercontent.com/json-schema-form/angular-schema-form-bootstrap/develop/dist/angular-schema-form-bootstrap.js

**Comment by warmy1004** [comment]
@Anthropic I tested it with the new version of bootstrap. The above issue is gone. But, it's still not working as expected. 

plnkr: https://plnkr.co/edit/GpmbaMPZmtgofOdV0Fye?p=preview

how to reproduce it :
case 1)
1. select 'no' for the parent question
2. then, click a 'validate' button

case 2) 
1. select 'yes' for the parent question
2. do not answer any conditional question and just click a validate button
3. change the parent question's answer from 'yes' to 'no' in order to hide the conditional questions

 -> for both cases, a required error message is still displayed. To me, it's also similar behavior which I saw above-  validation is always recursively and automatically happening for the entire form including not displayed things on the screen. 

If you have anything else, let me know please.

**Comment by warmy1004** [comment]
@Anthropic Also, I found another thing: validation is failing. You can use the same plnkr which I posted right above. 

how to reproduce it:
case 3) 
1. select 'yes' for the parent question
2. answer all three conditional questions. Especially, you need to select 'NO' for the last child question (Do you like animals) 
3. click a 'validate' button 

-> you will see a required button from the last child question.  What the ideal case is it should not be in error state because I selected 'no'. 

The thing is this last radio button question's type was set to be boolean instead of string. In this case is failing the validation. However, if I change radio buttons' type to 'string', then validation is working as I expected. Of course, case 1 and case 2 are not issue anymore if I change the parent radio button type's from boolean to string. 
This means i can't use boolean type of radio buttons through schema-form. Is this right?

**Comment by warmy1004** [comment]
@Anthropic sorry for multiple messages, but I want to let you know one more thing about checkbox

To show this, I slightly updated my plnkr : https://plnkr.co/edit/GpmbaMPZmtgofOdV0Fye?p=preview
[NOTE] 
(1) I added a string type of radio button as one of children questions, so you can easily see what the difference is. 
(2) I added a checkbox question as a child question, so you can reproduce the issue I found following the steps below. 

how to reproduce it :
1. select 'yes' for the parent question
2. check the checkbox question
3. change the parent question's answer from 'yes' to 'no'
4. change the parent question's answer again to 'yes' in order to display children questions
-> then check box question is displayed as a success state instead of a normal state.

**Comment by Anthropic** [comment]
@warmy1004 thanks for the info I will try to get to it soon, been sick for two weeks now, surviving on Codral... to build the decorator if you have changed templates it uses webpack `npm run build` or `npm run dist` to get minified, those commands will generate the dist folder in any of the three repo although core is unminified, webpack pulls in the templates via **import** statements from https://github.com/json-schema-form/angular-schema-form-bootstrap/blob/develop/src/bootstrap-decorator.js#L2

The variable generated then gets added as they used to by **definedDecorator** https://github.com/json-schema-form/angular-schema-form-bootstrap/blob/develop/src/bootstrap-decorator.js#L92

Probably better to add the templates as **add-ons** instead to future proof their addition so you don't need to modify anything in the repo, but that's how you do it if you need to or have your own decorator to work with. I will make sure to add that to the documentation by the time I get to beta.

**Comment by Anthropic** [comment]
@warmy1004 I've got your cases working locally, try the latest commit, the only issue it still has is that selecting yes then no then yes leaves the checkbox showing a required error. But I am not sure what to do with that, technically it is accurate. What we're missing is support for `const`.

Anyway feel free to give it a go, I'm going to close this issue as I believe most items are covered now, please make a new issue if there are any further issues.
